src/sql/workbench/browser/modal/optionsDialog.ts

```diff
@@ -107,7 +107,8 @@ export class OptionsDialog extends Modal {
 			attachButtonStyler(this.backButton, this._themeService, { buttonBackground: SIDE_BAR_BACKGROUND, buttonHoverBackground: SIDE_BAR_BACKGROUND });
 		}
 		let okButton = this.addFooterButton(localize('optionsDialog.ok', 'OK'), () => this.ok());
-		let closeButton = this.addFooterButton(this.options.cancelLabel || localize('optionsDialog.cancel', 'Cancel'), () => this.cancel());
+		let cancelLabel = this.options && this.options.cancelLabel ? this.options.cancelLabel :  localize('optionsDialog.cancel', 'Cancel');
```

Comment:
Bases on the signature of the constructor options is not a optional field, therefore it should always exist. This sounds like an issue with who ever is constructing the dialog, not the dialog itself.

Reply:
I think providing a default label for the Cancel button seems reasonable.  Is there any reason we wouldn't want to do this?  

Regarding `options` being required, a couple notes (1) we don't check that the options provided in the constructor is not null so presumably it could be passed as null (2) we don't check `cancelLabel` either, so presumably it could be passed as null as well even if `options` is provided.  I think a good practice would be to have the constructor throw an error if the provided input isn't valid.

Reply:
Why does the options need to be mandatory ?

Reply:
Also the option in the constructor of type  IOptionsDialogOptions is different what the open method accepts, which is type of ServiceOption[]. This is internallyc alling render. So fundamental IOptionsDialogOptions  should have a default variance. So I dont believe construction of this dialog is a problem

Reply:
I don't know why options is mandatory, it probably has to do with how the contructor decorators work but if we had strict-null-checking active, it would throw something like
`null is not assignable to type IOptionsDialogOptions`

Reply:
which I dont think it is doing currently as its null\undefined

Reply:
Even if we want options to be required (which we don't currently verify in the code), I don't think we want `cancelLabel` to be a required property?  I don't have an issue with this PR other than it is revealing a design problem with the constructor parameters, as far as I can tell.

Is there a recommended alternative here?  I don't think requiring the caller to provider a `cancelLabel` is the correct solution.

Reply:
`cancelLabel` is already not required per `this.options.cancelLabel || localize('optionsDialog.cancel', 'Cancel')`.

The error here is that `this.options` is undefined not `this.options.cancelLabel` is undefined.

Theres no problem with this solution. I'm just stating in the future if we enable `strict-null-checking` (which would catch this exact issue for us), this wouldn't be the solution that would solve the problem. Instead the solution would be to change the `undefined` being past here
https://github.com/Microsoft/azuredatastudio/blob/master/src/sql/workbench/services/backup/browser/backupUiService.ts#L73
to `{}` instead.

Reply:
The reason being that the actual parameter `options` is not optional (per the constructor signature) but every field in the object is optional.

Reply:
If you look at the call to `super`, `options` is optional there, so it probably should be here as well. This fix is fine to me though I agree with Anthony that it won't hold up when we move to strict null checks. But until we do support strict null checks I think it's best to assume that `options` could be undefined. This is a great example of why we should adopt strict null checks like VS Code

Reply:
> `cancelLabel` is already not required per `this.options.cancelLabel || localize('optionsDialog.cancel', 'Cancel')`.
> 
> The error here is that `this.options` is undefined not `this.options.cancelLabel` is undefined.
> 
> Theres no problem with this solution. I'm just stating in the future if we enable `strict-null-checking` (which would catch this exact issue for us), this wouldn't be the solution that would solve the problem. Instead the solution would be to change the `undefined` being past here
> https://github.com/Microsoft/azuredatastudio/blob/master/src/sql/workbench/services/backup/browser/backupUiService.ts#L73
> to `{}` instead.

This doesn't feel right. It would be like having consumers initialize and send object they dont care about right ? Can we make this optional for remove this from constructor and accept it as part of show/open/render ?

Reply:
Hmm, I guess we can't make `options` optional due to the way the `instanatiationSerivce` works?  In this case, I'd personally prefer the constructor to do the nullcheck and set options itself to `{}`.  

For the API consumer I think it's more natural to pass `undefined` then `{}` for parameters that you don't care about. 
 Ideally, the parameter would not be required if it really isn't required, though I guess either of the other suggested alternatives is fine.

Reply:
@MattIrv , @anthonydresser , @kburtram , I am thinking of getting rid of options as state from OptionsDialog as the base class already maintains this as _modelOptions and defaults it like below:

this._modalOptions = options || Object.create(null);
mixin(this._modalOptions, defaultOptions, false);

I am thinking of making _modalOptions as protected and then accessing this from the child class.  I dont think its a good idea having two states for same thing. With this we dont have to initialize it separately.

I will share PR, let me know what you guys think